crates/goose/src/agents/truncate.rs

```diff
@@ -246,152 +101,170 @@ impl Agent for TruncateAgent {
-        is_tool_use: bool,
-    ) -> Vec<Message> {
-        let mut conversation = Vec::<Message>::new();
-
-        if is_tool_use {
-            (0..interactions_count).for_each(|i| {
-                let tool_output = format!("{:?}{}", SMALL_MESSAGE, i);
-                conversation.push(
-                    Message::user()
-                        .with_tool_response("id:0", Ok(vec![Content::text(tool_output)])),
-                );
-                conversation.push(Message::assistant().with_text(format!(
-                    "{:?}{}",
-                    SMALL_MESSAGE,
-                    i + 1
-                )));
-            });
-        } else {
-            (0..interactions_count).for_each(|i| {
-                conversation.push(Message::user().with_text(format!("{:?}{}", SMALL_MESSAGE, i)));
-                conversation.push(Message::assistant().with_text(format!(
-                    "{:?}{}",
-                    SMALL_MESSAGE,
-                    i + 1
-                )));
-            });
+        // Set the user_message field in the span instead of creating a new event
+        if let Some(content) = messages
+            .last()
+            .and_then(|msg| msg.content.first())
+            .and_then(|c| c.as_text())
+        {
+            debug!("user_message" = &content);
         }
 
-        conversation
-    }
-    #[tokio::test]
-    async fn test_simple_conversation_no_truncation() -> anyhow::Result<()> {
-        let conversation = create_basic_valid_conversation(1, false);
-        let messages = call_enforce_ctx_limit(&conversation).await?;
-        assert_eq!(messages.len(), conversation.len());
-        Ok(())
-    }
-    #[tokio::test]
-    async fn test_truncation_when_conversation_history_too_big() -> anyhow::Result<()> {
-        let conversation = create_basic_valid_conversation(5000, false);
-        let messages = call_enforce_ctx_limit(&*conversation).await?;
-        assert_eq!(conversation.len() > messages.len(), true);
-        assert_eq!(messages.len() > 0, true);
-        Ok(())
-    }
-
-    #[tokio::test]
-    async fn test_truncation_when_single_user_message_too_big() -> anyhow::Result<()> {
-        let oversized_message: String = iter::repeat(SMALL_MESSAGE)
-            .take(10000)
-            .collect::<Vec<&str>>()
-            .join("");
-        let mut conversation = create_basic_valid_conversation(3, false);
-        conversation.push(Message::user().with_text(oversized_message));
-
-        let messages = call_enforce_ctx_limit(&*conversation).await;
+        Ok(Box::pin(async_stream::try_stream! {
+            let _reply_guard = reply_span.enter();
+            loop {
+                // Attempt to get completion from provider
+                match capabilities.provider().complete(
+                    &system_prompt,
+                    &messages,
+                    &tools,
+                ).await {
+                    Ok((response, usage)) => {
```

Comment:
i think we want to reset truncation attempts to 0 after a successful inference (we succeed truncating, but it could come up again if the tool use loop goes on a long while)